lightning/src/ln/channel.rs

```diff
@@ -2518,6 +2523,52 @@ struct CommitmentTxInfoCached {
 	feerate: u32,
 }
 
+/// Contents of a wire message that fails an HTLC backwards. Useful for [`Channel::fail_htlc`] to
+/// fail with either [`msgs::UpdateFailMalformedHTLC`] or [`msgs::UpdateFailHTLC`] as needed.
+pub(super) trait FailHTLCContents {
+	type Message;
+	fn message(self, htlc_id: u64, channel_id: ChannelId) -> Self::Message;
+	fn message_name() -> &'static str;
+	fn inbound_htlc_state(self) -> InboundHTLCState;
+	fn htlc_update_awaiting_ack(self, htlc_id: u64) -> HTLCUpdateAwaitingACK;
```

Comment:
Somewhat pedantic, but:
- naming could use `to_` prefix for methods taking `self`
- `message_name` arguably belongs on it's own trait (`MessageName`?) where the `Message` associated type has a trait bound on it.